Replace web installer by CLI task, refs #13452

jraddaoui opened the pull request:
- Remove sfInstallPlugin and move sfInstall class to arInstall.
- Tweak arInstall to throw exceptions instead of returning errors.
- Move DB configuration handler class into arInstall.
- Remove install layout and related styles.
- Create install task to configure and initialize a new instance.
- Normalize purge and install tasks.

Other fixes:

- Fix QubitActor index on save check.
- Allow Propel's default DB name change.
- Install apcu extension in Docker.
- Avoid DB exceptions on routing initialization.
- Avoid PHP notice.

jraddaoui commented:
More details in:

https://github.com/artefactual/atom-docs/blob/2.7/admin-manual/installation/linux/ubuntu.rst#run-the-installer
https://github.com/artefactual/atom-docs/blob/2.7/admin-manual/maintenance/cli-tools.rst#installer

sbreker commented on lib/model/QubitObject.php:
```diff
@@ -35,7 +35,10 @@ public function __isset($name)
                     $row = $statement->fetch();
-                    $this->values['slug'] = $row[0];
+
+                    if (false !== $row = $statement->fetch()) {
```
Is line 37 redundant?  On line 39 `$row` appears to be immediately overwritten by a second fetch().

jraddaoui replied:
Right, good catch!

jraddaoui replied:
Addressed in 684e29b.

sbreker commented on lib/routing/QubitMetadataRoute.class.php:
```diff
@@ -77,17 +77,8 @@ public function matchesUrl($url, $context = [])
-
-                return $parameters;
+            if (null === $this->resource = QubitObject::get($criteria)->__get(0)) {
+                return false;
```
What happens in this case? 500 error if arInstall has not completed?

jraddaoui replied:
I have not tested it, but that was my assumption. I'll give it a try.

jraddaoui replied:
Good thing you pointed this out as it's quite confusing. Based on the URL you hit, when you go to the site without running the installer you get different errors ...

- Using a slug you get our normal 500 error page and the following error in the logs:

```
[11-Jun-2021 17:27:49] WARNING: [pool atom] child 16 said into stderr: "NOTICE: PHP message: No connection information in your runtime configuration file for datasource [propel]"
```

- Without slug you get a default 500 error page and a bit more confusing log:

```
[11-Jun-2021 17:21:02] WARNING: [pool atom] child 16 said into stderr: "NOTICE: PHP message: Empty module and/or action after parsing the URL "/" (/)."
[11-Jun-2021 17:21:02] WARNING: [pool atom] child 16 said into stderr: "NOTICE: PHP message: PHP Fatal error:  Uncaught sfError404Exception: Action "default/error404" does not exist. in /atom/src/cache/qubit/prod/config/config_core_compile.yml.php:660"
[11-Jun-2021 17:21:02] WARNING: [pool atom] child 16 said into stderr: "Stack trace:"
[11-Jun-2021 17:21:02] WARNING: [pool atom] child 16 said into stderr: "#0 /atom/src/vendor/symfony/lib/exception/sfError404Exception.class.php(49): sfController->forward('default', 'error404')"
[11-Jun-2021 17:21:02] WARNING: [pool atom] child 16 said into stderr: "#1 /atom/src/cache/qubit/prod/config/config_core_compile.yml.php(2396): sfError404Exception->printStackTrace()"
[11-Jun-2021 17:21:02] WARNING: [pool atom] child 16 said into stderr: "#2 /atom/src/vendor/symfony/lib/util/sfContext.class.php(170): sfFrontWebController->dispatch()"
[11-Jun-2021 17:21:02] WARNING: [pool atom] child 16 said into stderr: "#3 /atom/src/index.php(6): sfContext->dispatch()"
[11-Jun-2021 17:21:02] WARNING: [pool atom] child 16 said into stderr: "#4 {main}"
[11-Jun-2021 17:21:02] WARNING: [pool atom] child 16 said into stderr: "  thrown in /atom/src/cache/qubit/prod/config/config_core_compile.yml.php on line 660"
```

I'm trying to normalize this,  but it's not as easy as it looks ... thoughts?

sbreker replied:
Previously we redirected to sfInstall when things weren't set up. Could we detect if the installer has not been run and  redirect to a static "Go run the AtoM CLI installer" page for all routes?

jraddaoui replied:
Ok, still a normal error page but now the error will be the same for all routes. In any case, lets hope they don't forget to run the installer.

jraddaoui replied:
Addressed in fa2a1dd and 2df334e.

sbreker commented on lib/arInstall.class.php:
```diff
@@ -0,0 +1,421 @@
+        $setting = new QubitSetting();
+        $setting->name = $name;
+        $setting->value = $value;
+        $setting->save();
```
Nitpick - insert newline before 419.

jraddaoui replied:
Why?

sbreker replied:
Please disregard! I believe I have misinterpreted the PHP-CS-Fixer rule.

sbreker commented on lib/arInstall.class.php:
```diff
@@ -0,0 +1,421 @@
+            mkdir(sfConfig::get('sf_log_dir'), 0777, true);
+        }
+
+        Qubit::createUploadDirsIfNeeded();
```
Nitpick: I think our coding standard states that there should be a blank line preceeding the last line of a function/method. This file has this issue in a number of places.

sbreker commented on lib/arInstall.class.php:
```diff
@@ -0,0 +1,421 @@
+            bin2hex(openssl_random_pseudo_bytes(16)),
+            $content
+        );
+        if (false === file_put_contents($settingsYmlPath, $content)) {
```
Nitpick: I believe coding standard states to have a newline following a multi-line statement. There are a few examples of this in this file.

jraddaoui replied:
I don't think that's the case. Or do you mean before control statements? In which case, `if` it's not included in PphCsFixer ruleset:

https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/3.0/doc/rules/whitespace/blank_line_before_statement.rst#rule-sets

sbreker replied:
Thanks for the link - Which ruleset are we using?

jraddaoui replied:
We use the `@PhpCsFixer` ruleset with a single exception to avoid inconsistencies in templates:

https://github.com/artefactual/atom/blob/qa/2.x/.php_cs#L24-L25

sbreker commented on lib/arInstall.class.php:
```diff
@@ -0,0 +1,421 @@
+        }
+
+        Qubit::createUploadDirsIfNeeded();
+        Qubit::createDownloadsDirIfNeeded();
```
Nitpick: AtoM's coding standard states that there should be a blank line before the last line of a function/method.

jraddaoui replied:
I think that's only if the function ends in a return or other control statements from the ones mentioned in the rule from the previous comment.

sbreker commented on lib/arInstall.class.php:
```diff
@@ -0,0 +1,421 @@
+        $databasesYmlPath = sfConfig::get('sf_config_dir').'/databases.yml';
+
+        // Read databases.yml contents from existing databases.yml,
+        // databases.yml.tmpl (for a Subversion checkout), or symfony
```
Subversion

jraddaoui replied:
Not sure what do you mean here.

sbreker replied:
The comment is out of date ('subversion' instead of 'github') - sorry I should have clarified!  😄

jraddaoui replied:
Maybe just `git`, guessing it could be checked from other repository ... 4997828.

sbreker commented on lib/arInstall.class.php:
```diff
@@ -0,0 +1,421 @@
+    {
+        $file = sfConfig::get('sf_root_dir').'/vendor/symfony/data/web/sf';
+        $link = sfConfig::get('sf_root_dir').'/sf';
+        if (!file_exists($link) && false === symlink($file, $link)) {
```
Newline before if?

jraddaoui replied:
As said before, not actually considered in the control statements that require an empty line before.

sbreker commented on lib/arInstall.class.php:
```diff
@@ -0,0 +1,421 @@
+            date('Y/m/d H:i:s'),
+            var_export($config, true)
+        );
+        if (false === file_put_contents($configFile, $content)) {
```
Newline before if.

jraddaoui replied:
Same.

sbreker commented on lib/task/tools/installTask.class.php:
```diff
@@ -0,0 +1,536 @@
+                'ERROR'
+            );
+
+            exit(2);
```
Is there a meaningful difference between `exit(1)` used below and `exit(2)` used above?

jraddaoui replied:
There are different exit/error codes, not something that we'd use but the general idea would be ...

- 0 = success
- 1 = aborted
- 2 = error

sbreker replied:
Opinion: These should be constants if there is meaning behind them - e.g. `exit(self::ERROR)`

jraddaoui replied:
I like that, thanks Steve.

jraddaoui replied:
I'm thinking this doesn't add a lot of value considering that the error code could be useful in automated process and the difference between 1 and 2 would require user interaction to abort the task. So I think just `exit(1);` in all cases should be enough.

jraddaoui replied:
Addressed in 0252ed5.

sbreker commented on lib/task/tools/installTask.class.php:
```diff
@@ -0,0 +1,536 @@
+        );
+
+        $this->logSection('install', 'Installation completed');
+    }
```
`execute()` is > 300 lines in length. Would readability be improved if it were broken up?

jraddaoui replied:
One of the downsides of being careful with the 80 chars line length and 4 chars indentation ... I try not to move things out unless they are "reused" somewhere else but I see your point. There is also some duplication between the install and purge task after configuration, so I'll see if we can avoid/improve that.

jraddaoui replied:
Addressed in 40ee253 and 9ca9016.

I hope that makes it a bit clearer. We could go a bit further extending the installTask from purgeTask, but it will require to change one of the task option names and the user input process to normalize them.

sbreker replied:
This looks great!

sbreker reviewed (changes requested):
This looks really great @jraddaoui - thanks!  I've noted a few small items to review.  

This will be a big improvement - looking forward to no more installation timeouts!  👍

sbreker reviewed (approved):
Looks great @jraddaoui!  👍

jraddaoui committed:
Replace web installer by CLI task, refs #13452

- Remove sfInstallPlugin and move sfInstall class to arInstall.
- Tweak arInstall to throw exceptions instead of returning errors.
- Move DB configuration handler class into arInstall.
- Remove install layout and related styles.
- Create install task to configure and initialize a new instance.
- Normalize purge and install tasks.
- Use new install task on CI.

Other fixes:

- Fix QubitActor index on save check.
- Allow Propel's default DB name change.
- Install apcu extension in Docker.
- Avoid DB exceptions on routing initialization.
- Avoid PHP notice.

Squashed commit of the following:

commit 9ca90160be6fd0e48f0f5b02d776cb2f06ff38b3
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon Jun 14 18:23:32 2021 +0200

    Extend installTask from purgeTask

    Allows to reuse `initializeDbAndEs` to avoid duplication.

commit 40ee2530dbaebf59f8669ccab95753b9064ac59b
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon Jun 14 18:21:38 2021 +0200

    Split install task and use task name in logs

commit 0252ed5e4d0f4cce8301a8a024a5675eec1d11c3
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon Jun 14 17:23:48 2021 +0200

    Use `exit(1);` in all cases

commit 499782857939d3130ad1532d4dbc3d48d6ee9a79
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Jun 11 23:39:09 2021 +0200

    Use git instead of Subversion

commit 684e29b581b1dfc260b65b563f85987643409e0f
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Jun 11 23:27:42 2021 +0200

    Avoid double fetch

commit 2df334e2a4c6ce490a1798b0a1a7d20ed20e1a69
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Jun 11 23:27:23 2021 +0200

    Avoid propel exceptions on routing init

    Based on a config setting that is only used from the install task.

commit fa2a1dd182f81e607307d9c72481b72304b97422
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Jun 11 23:25:37 2021 +0200

    Revert "Avoid DB exceptions on routing initialization"

    This reverts commit ac8a9d3d42a71afa4a0c607c41e3233f6d692ef8.

commit a7e776198805f2e17d33edea546bc7057d4abab2
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Thu Jun 10 07:06:37 2021 +0200

    Use CLI installer in CI

commit 5d72eb7bd1d919af3a288e0fcd55cbba08cb24ed
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Wed Jun 9 16:52:16 2021 +0200

    Replace web installer by CLI task, refs #13452

    - Remove sfInstallPlugin and move sfInstall class to arInstall.
    - Tweak arInstall to throw exceptions instead of returning errors.
    - Move DB configuration handler class into arInstall.
    - Remove install layout and related styles.
    - Create install task to configure and initialize a new instance.
    - Normalize purge and install tasks.

commit 04979bafdceede60fbbce1e914f475512e55e1a7
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Thu May 6 17:37:40 2021 +0200

    Avoid PHP notice

    Notice: Trying to access array offset on value of type bool.

commit 81d2b171922c5490266e636c05b931a8c8be7c4e
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Thu May 6 17:31:43 2021 +0200

    Avoid DB exceptions on routing initialization

commit 394cd5cba2035518b0c7ff7bde6f82ba35164cae
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon May 24 16:33:14 2021 +0200

    Install apcu extension in Docker

commit 8b04c2d2bae3619334fa9d5589fcd9889940bc40
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon May 24 16:44:25 2021 +0200

    Allow Propel's default DB name change

commit 136980092ba707729f0cc249da5456b54df4a4df
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon May 3 14:29:08 2021 +0200

    Fix QubitActor index on save check

    Check `indexOnSave` property before updating the resource and relations.

jraddaoui commented:
Thanks @sbreker!
